gprofiler/spark/spark_collector.py

```diff
@@ -0,0 +1,822 @@
+                url = self._join_url_dir(url, directory)
+
+        logger.debug(f"Spark check URL: {url}")
+        response = requests.get(url, params={k: v for k, v in kwargs.items() if v is not None}, timeout=3)
+        response.raise_for_status()
+        return response
+
+    def _rest_request_to_json(self, address: str, object_path: str, *args: Any, **kwargs: Any) -> Any:
+        """
+        Query the given URL and return the JSON response
+        """
+        return self._rest_request(address, object_path, *args, **kwargs).json()
+
+    def _get_request_url(self, url: str) -> str:
+        """
+        Get the request address, build with proxy if necessary
+        """
+        parsed = urlparse(url)
+
+        _url = url
+        if not (parsed.netloc and parsed.scheme):
+            _url = urljoin(self._master_address, parsed.path)
+
+        return _url
+
+    @staticmethod
+    def _join_url_dir(url: str, *args: Any) -> str:
+        """
+        Join a URL with multiple directories
+        """
+        for path in args:
+            url = url.rstrip("/") + "/"
+            url = urljoin(url, path.lstrip("/"))
+
+        return url
+
+
+class SparkSampler(object):
+    """
+    Spark cluster metrics sampler
+    """
+
+    def __init__(
+        self,
+        sample_period: float,
+        storage_dir: str,
+        api_client: Optional[APIClient] = None,
+    ):
+        self._master_address: Optional[str] = None
+        self._spark_mode: Optional[str] = None
+        self._collection_thread: Optional[Thread] = None
+        self._sample_period = sample_period
+        self._stop_event = Event()
+        self._spark_sampler: Optional[SparkCollector] = None
+        self._stop_collection = False
+        self._is_running = False
+        self._storage_dir = storage_dir
+        if self._storage_dir is not None:
+            assert os.path.exists(self._storage_dir) and os.path.isdir(self._storage_dir)
+        else:
+            logger.debug("Output directory is None. Will add metrics to queue")
+        self._client = api_client
+
+    def _get_yarn_config_path(self, process: psutil.Process) -> str:
+        env = process.environ()
+        if "HADOOP_CONF_DIR" in env:
+            path = env["HADOOP_CONF_DIR"]
+            logger.debug("Found HADOOP_CONF_DIR variable", hadoop_conf_dir=path)
+        else:
+            path = "/etc/hadoop/conf/"
+            logger.info("Could not find HADOOP_CONF_DIR variable, using default path", hadoop_conf_dir=path)
+        return os.path.join(path, "yarn-site.xml")
+
+    def _get_yarn_config(self, process: psutil.Process) -> Optional[ET.Element]:
+        config_path = self._get_yarn_config_path(process)
+
+        logger.debug("Trying to open yarn config file for reading", config_path=config_path)
+        try:
+            # resolve config path against process' filesystem root
+            process_relative_config_path = resolve_host_path(process, self._get_yarn_config_path(process))
```

Comment:
`resolve_host_path` is a weird name because neither gProfiler not the process are necessarily on the host. Oh well. The function does seem to be the correct one for the use case :shrug: just bad naming